allennlp/modules/stacked_elmo_lstm.py

```diff
@@ -0,0 +1,146 @@
+            of the LSTM. Each tensor has shape (1, batch_size, output_dimension).
+
+        Returns
+        -------
+        output_sequence : ``torch.FloatTensor``
+            The encoded sequence of shape (num_layers, batch_size, sequence_length, hidden_size)
+        final_states: ``Tuple[torch.FloatTensor, torch.FloatTensor]``
+            The per-layer final (state, memory) states of the LSTM, with shape
+            (num_layers, batch_size, 2 * hidden_size) and  (num_layers, batch_size, 2 * cell_size)
+            respectively.
+        """
+        if not initial_state:
+            hidden_states = [None] * len(self.forward_layers)
+        elif initial_state[0].size()[0] != len(self.forward_layers):
+            raise ConfigurationError("Initial states were passed to forward() but the number of "
+                                     "initial states does not match the number of layers.")
+        else:
+            hidden_states = list(zip(initial_state[0].split(1, 0),
+                                     initial_state[1].split(1, 0)))
+
+        inputs, batch_lengths = pad_packed_sequence(inputs, batch_first=True)
+        forward_output_sequence = inputs
+        backward_output_sequence = inputs
+
+        final_states = []
+        sequence_outputs = []
+        for layer_index, (forward_layer, backward_layer, state) in enumerate(zip(self.forward_layers,
+                                                                                 self.backward_layers,
+                                                                                 hidden_states)):
+            forward_cache = forward_output_sequence
+            backward_cache = backward_output_sequence
+
+            forward_output_sequence, forward_state = forward_layer(forward_output_sequence,
+                                                                   batch_lengths,
+                                                                   state)
+            backward_output_sequence, backward_state = backward_layer(backward_output_sequence,
+                                                                      batch_lengths,
+                                                                      state)
+            # Vanilla highway layers.
+            if layer_index != 0:
+                forward_output_sequence += forward_cache
+                backward_output_sequence += backward_cache
+
+            sequence_outputs.append(torch.cat([forward_output_sequence,
+                                               backward_output_sequence], -1))
+            # Append the state tuples in a list, so that we can return
+            # the final states for all the layers.
+            final_states.append((torch.cat([forward_state[0], backward_state[0]], -1),
+                                 torch.cat([forward_state[1], backward_state[1]], -1)))
+        # TODO(Mark): figure out the best api to return these. We need to specify the
```

Comment:
We should move this logic elsewhere.

Reply:
I think it will be cleanest if we define a set of mixing functions elsewhere and pass one of them to this module, rather than having another layer after this.

Reply:
There is other functionality that we need to add to make easy to use and work with ELMo -- e.g. removing `<S>` & `</S>`, loading the pretrained weights, making it stateful -- all before the weighting of layers can happen.  It also needs to be paired with the specific ELMo token embedder so I was planning to add those elsewhere in something ELMo specific.